tests/unit_tests/drivers/test_jobscript_driver.py

```diff
@@ -62,72 +160,121 @@ def fixture_job_options(tmp_path, job_id, experiment_name, injected_input_files)
     input_template_1, input_template_2 = input_templates
 
     driver = JobscriptDriver(
         parameters=parameters,
         jobscript_template="",
-        executable="",
+        executable=executable,
         input_templates={"input_1": input_template_1, "input_2": input_template_2},
     )
 
     return driver
 
 
-def test_jobscript_driver_multiple_input_files(
-    jobscript_driver, job_options, injected_input_files, parameters
+@pytest.fixture(name="args_init")
+def fixture_args_init(
+    parameters,
+    jobscript_template,
+    executable,
+    input_template,
+    files_to_copy,
+    data_processor,
+    gradient_data_processor,
+    jobscript_file_name,
+    extra_options,
 ):
+    """Arguments to initialize a JobscriptDriver.
+
+    These arguments are meant for initialization with the default
+    constructor.
+    """
+    # pylint: disable=unused-argument
+    args_init = locals()
+    args_init["input_templates"] = args_init["input_template"]
+    args_init.pop("input_template")
+    return args_init
+
+
+def assert_jobscript_driver_attributes(jobscript_driver, args_init):
+    """Assert that the jobscript driver attributes are set correctly."""
+    args_init["files_to_copy"].append(args_init["input_templates"])
+
+    assert jobscript_driver.parameters == args_init["parameters"]
+    assert jobscript_driver.input_templates == {"input_file": args_init["input_templates"]}
+    assert jobscript_driver.jobscript_template == args_init["jobscript_template"]
+    assert jobscript_driver.jobscript_options["executable"] == args_init["executable"]
+    assert jobscript_driver.files_to_copy == args_init["files_to_copy"]
+    assert jobscript_driver.data_processor == args_init["data_processor"]
+    assert jobscript_driver.gradient_data_processor == args_init["gradient_data_processor"]
+    assert jobscript_driver.jobscript_file_name == args_init["jobscript_file_name"]
+    assert jobscript_driver.jobscript_options.items() >= args_init["extra_options"].items()
+
+
+def assert_subprocess_error(
+    parameters, raise_error_on_jobscript_failure, jobscript_driver, job_options
+):
+    """Assert that jobscript driver run raises an error if so intended."""
+    sample_dict = parameters.sample_as_dict(np.array([1, 2]))
+    sample = np.array(list(sample_dict.values()))
+
+    if raise_error_on_jobscript_failure:
+        with pytest.raises(SubprocessError):
+            jobscript_driver.run(
+                sample=sample,
+                job_id=job_options.job_id,
+                num_procs=job_options.num_procs,
+                experiment_dir=job_options.experiment_dir,
+                experiment_name=job_options.experiment_name,
+            )
+    else:
+        jobscript_driver.run(
+            sample=sample,
+            job_id=job_options.job_id,
+            num_procs=job_options.num_procs,
+            experiment_dir=job_options.experiment_dir,
+            experiment_name=job_options.experiment_name,
+        )
```

Comment:
The code seems misplaced. Basically everything is done in this function instead of the tests itself. I get what you did here to reduce code. I did some weird coding myself to circumvent this problem in other tests. I looked for a different solution and stumbled upon [this](https://stackoverflow.com/a/74467561). I think this would be a nice feature that we could reuse in other tests to parameterize expected failure as well as successful execution. What do you think?

Reply:
That's a nice, clean way of testing expected failure and successful execution, thanks for looking into this!